internal/storage/prefixdb.go

```diff
@@ -0,0 +1,632 @@
+//
+// It implements the [Iterator] interface for type prefixDBIterator.
+func (it *prefixDBIterator) Close() error {
+	if err := it.source.Close(); err != nil {
+		return fmt.Errorf("closing prefixed DB namespace iterator: %w", err)
+	}
+
+	return nil
+}
+
+// Domain returns the start (inclusive) and end (exclusive) limits of the iterator.
+// Callers must not modify the returned slices. Instead, they should make a copy if
+// they need to modify them.
+//
+// It implements the [Iterator] interface for type prefixDBIterator.
+func (it *prefixDBIterator) Domain() (start []byte, end []byte) {
+	return it.start, it.end
+}
+
+// Error returns the last error encountered by the iterator, if any.
+//
+// It implements the [Iterator] interface for type prefixDBIterator.
+func (it *prefixDBIterator) Error() error {
+	if err := it.source.Error(); err != nil {
+		return err
+	}
+	return it.err
+}
+
+// Key returns the key, stripped of the prefix, at the current position or panics if
+// the iterator is invalid.
+// The caller should not modify the contents of the returned slice and its contents
+// may change on the next call to [Next]. Therefore, callers should make a copy of
+// the returned slice if they need to modify it.
+//
+// It implements the [Iterator] interface for type prefixDBIterator.
+func (it *prefixDBIterator) Key() []byte {
+	it.assertIsValid()
+	key := it.source.Key()
+
+	// we have checked that the key is valid in the call to assertIsValid()
+	return key[len(it.prefix):]
+}
+
+// Next moves the iterator to the next key in the database, as defined by order
+// of iteration, or panics if the iterator is invalid.
+//
+// It implements the [Iterator] interface for type prefixDBIterator.
+func (it *prefixDBIterator) Next() {
+	it.assertIsValid()
+	it.source.Next()
+
+	var (
+		srcItInvalid  = !it.source.Valid()
+		missingPrefix = !bytes.HasPrefix(it.source.Key(), it.prefix)
+	)
+	if srcItInvalid || missingPrefix {
+		it.valid = false
+		return
+	}
+
+	if bytes.Equal(it.source.Key(), it.prefix) {
+		// Empty keys are not allowed, so if a key exists in the database that
+		// exactly matches the prefix we need to skip it.
+		// The key is going to be lexicograpically smaller than the first "correct"
+		// key of the form prefix+key, because
+		// prefix < prefix+key, e.g., "foo" < "fooa".
+		// Therefore, we only need to skip it to position the iterator at the first
+		// "correct" key.
+		it.Next()
+	}
+}
+
+// Valid returns whether the current iterator is valid. Once invalid, the Iterator
+// remains invalid forever.
+//
+// It implements the [Iterator] interface for type prefixDBIterator.
+func (it *prefixDBIterator) Valid() bool {
+	if !it.valid || it.err != nil || !it.source.Valid() {
+		it.valid = false
+		return false
+	}
+
+	var (
+		key            = it.source.Key()
+		keyTooShort    = len(key) < len(it.prefix)
+		prefixMismatch = !bytes.Equal(key[:len(it.prefix)], it.prefix)
+	)
+	if keyTooShort || prefixMismatch {
```

Comment:
See my comment about the `Next` method above. The same considerations apply here.